File: back/app/domain/business_flow/bpmn_semantic.py
"""Type-specific business semantics for non-Task BPMN elements."""

from __future__ import annotations

from typing import Any, Mapping

SCHEMA_VERSION = 1
DATA_ELEMENT_TYPES = {"DATA_OBJECT", "DATA_INPUT", "DATA_OUTPUT", "DATA_STORE"}

NODE_SEMANTIC_TYPES = {
    ("EVENT", "START", None, None): "startEvent",
    ("EVENT", "INTERMEDIATE", None, None): "intermediateEvent",
    ("EVENT", "END", None, None): "endEvent",
    ("SUB_PROCESS", None, None, "TRANSACTION"): "transaction",
    ("GATEWAY", None, "EXCLUSIVE", None): "exclusiveGateway",
    ("GATEWAY", None, "INCLUSIVE", None): "inclusiveGateway",
    ("GATEWAY", None, "PARALLEL", None): "parallelGateway",
    ("GATEWAY", None, "COMPLEX", None): "complexGateway",
    ("DATA_OBJECT", None, None, None): "dataObject",
    ("DATA_INPUT", None, None, None): "dataInput",
    ("DATA_OUTPUT", None, None, None): "dataOutput",
    ("DATA_STORE", None, None, None): "dataStore",
}
EDGE_SEMANTIC_TYPES = {
    "SEQUENCE": "sequenceFlow",
    "MESSAGE": "messageFlow",
    "ASSOCIATION": "association",
}
NAME_KEYS = {
    "startEvent": "eventName",
    "intermediateEvent": "eventName",
    "endEvent": "eventName",
    "transaction": "transactionName",
    "exclusiveGateway": "decisionName",
    "inclusiveGateway": "decisionName",
    "parallelGateway": "gatewayName",
    "complexGateway": "gatewayName",
    "dataObject": "dataName",
    "dataInput": "dataName",
    "dataOutput": "dataName",
    "dataStore": "dataName",
    "sequenceFlow": "flowName",
    "messageFlow": "messageName",
    "association": "associationName",
}

FIELD_KINDS: dict[str, dict[str, str]] = {
    "startEvent": {
        "eventName": "text", "triggerType": "text", "triggerSource": "text",
        "startCondition": "text", "inputDataRefs": "list", "initiator": "text",
        "frequency": "text", "preCheckRules": "list",
    },
    "intermediateEvent": {
        "eventName": "text", "catchOrThrow": "text", "eventDefinition": "text",
        "interrupting": "bool", "timeout": "text", "messageName": "text",
        "errorCode": "text", "escalationCode": "text", "businessMeaning": "text",
    },
    "endEvent": {
        "eventName": "text", "resultType": "text", "finalBusinessState": "text",
        "outputDataRefs": "list", "notifyTargets": "list", "auditRequired": "bool",
        "rollbackRequired": "bool",
    },
    "transaction": {
        "transactionName": "text", "transactionType": "text",
        "successCriteria": "list", "cancelTriggers": "list",
        "compensationPolicy": "text", "compensationOrder": "text",
        "consistencyLevel": "text", "timeout": "text", "isolationNote": "text",
        "compensationTasks": "list", "partialSuccessPolicy": "text",
        "auditRequired": "bool",
    },
    "exclusiveGateway": {
        "decisionName": "text", "decisionVariable": "text", "branches": "objects",
        "defaultFlowId": "text", "conditionExpressionType": "text",
        "mutuallyExclusive": "bool", "coverageRequired": "bool",
    },
    "inclusiveGateway": {
        "decisionName": "text", "branchConditions": "objects",
        "allowMultipleBranches": "bool", "mergePolicy": "text",
        "minSelectedBranches": "int", "coverageRequired": "bool",
    },
    "parallelGateway": {
        "gatewayName": "text", "parallelMode": "text", "waitForAll": "bool",
        "expectedBranches": "list", "partialFailurePolicy": "text",
        "timeout": "text", "concurrencyLimit": "int",
    },
    "complexGateway": {
        "gatewayName": "text", "activationCondition": "text",
        "completionCondition": "text", "requiredCount": "int", "totalCount": "int",
        "customRule": "text", "explanation": "text",
    },
    "dataObject": {
        "dataName": "text", "entityName": "text", "schemaRef": "text",
        "lifecycleState": "text", "ownerActivityRef": "text",
        "readByRefs": "list", "writeByRefs": "list",
    },
    "dataInput": {
        "dataName": "text", "sourceType": "text", "sourceRef": "text",
        "required": "bool", "validationRules": "list", "exampleValue": "text",
        "sensitiveLevel": "text", "defaultValue": "text",
    },
    "dataOutput": {
        "dataName": "text", "targetType": "text", "targetRef": "text",
        "outputContract": "text", "transformRule": "text",
        "successOutput": "text", "failureOutput": "text",
    },
    "dataStore": {
        "dataName": "text", "storeType": "text", "systemRef": "text",
        "accessMode": "text", "consistencyLevel": "text",
        "retentionPolicy": "text", "privacyLevel": "text",
    },
    "sequenceFlow": {
        "flowName": "text", "flowKind": "text", "conditionText": "text",
        "conditionExpression": "text", "conditionExpressionType": "text",
        "priority": "int", "isDefault": "bool", "businessRuleRefs": "list",
        "testScenarioType": "text", "expectedResult": "text",
    },
    "messageFlow": {
        "messageName": "text", "businessMeaning": "text", "senderRef": "text",
        "receiverRef": "text", "payloadDataRefs": "list", "deliveryMode": "text",
        "timeout": "text", "testScenarioType": "text", "expectedResult": "text",
    },
    "association": {
        "associationName": "text", "businessMeaning": "text", "direction": "text",
        "dataRole": "text", "testScenarioType": "text", "expectedResult": "text",
    },
}

DEFAULT_VALUES: dict[str, dict[str, Any]] = {
    "startEvent": {"triggerType": "manual"},
    "intermediateEvent": {
        "catchOrThrow": "catch", "eventDefinition": "message", "interrupting": False,
    },
    "endEvent": {"resultType": "success", "auditRequired": False, "rollbackRequired": False},
    "transaction": {
        "transactionType": "business", "compensationOrder": "reverseOrder",
        "consistencyLevel": "eventual", "auditRequired": False,
    },
    "exclusiveGateway": {
        "conditionExpressionType": "naturalLanguage", "mutuallyExclusive": True,
        "coverageRequired": True,
    },
    "inclusiveGateway": {
        "allowMultipleBranches": True, "minSelectedBranches": 1,
        "coverageRequired": True,
    },
    "parallelGateway": {"parallelMode": "forkJoin", "waitForAll": True},
    "dataInput": {"sourceType": "user", "required": False, "sensitiveLevel": "normal"},
    "dataOutput": {"targetType": "ui"},
    "dataStore": {
        "storeType": "database", "accessMode": "readWrite",
        "consistencyLevel": "eventual", "privacyLevel": "internal",
    },
    "sequenceFlow": {
        "flowKind": "normal", "conditionExpressionType": "naturalLanguage",
        "isDefault": False, "testScenarioType": "normal",
    },
    "messageFlow": {"deliveryMode": "async", "testScenarioType": "normal"},
    "association": {
        "direction": "none", "dataRole": "reference", "testScenarioType": "normal",
    },
}

ENUM_VALUES: dict[str, dict[str, set[str]]] = {
    "startEvent": {"triggerType": {"manual", "message", "timer", "signal", "condition"}},
    "intermediateEvent": {
        "catchOrThrow": {"catch", "throw"},
        "eventDefinition": {"message", "timer", "error", "signal", "compensation", "escalation"},
    },
    "endEvent": {
        "resultType": {"success", "failure", "cancel", "terminate", "error", "compensation"},
    },
    "transaction": {
        "transactionType": {"technical", "business", "saga"},
        "compensationOrder": {"reverseOrder", "forwardOrder", "custom"},
        "consistencyLevel": {"strong", "eventual"},
    },
    "exclusiveGateway": {
        "conditionExpressionType": {"naturalLanguage", "feel", "javascript", "sql"},
    },
    "parallelGateway": {"parallelMode": {"fork", "join", "forkJoin"}},
    "dataInput": {
        "sourceType": {"user", "system", "api", "message", "file"},
        "sensitiveLevel": {"normal", "internal", "confidential"},
    },
    "dataOutput": {"targetType": {"api", "database", "message", "file", "ui"}},
    "dataStore": {
        "storeType": {"database", "cache", "file", "mq", "externalSystem"},
        "accessMode": {"read", "write", "readWrite"},
        "consistencyLevel": {"strong", "eventual"},
        "privacyLevel": {"normal", "internal", "confidential"},
    },
    "sequenceFlow": {
        "flowKind": {"normal", "conditional", "default", "exception", "compensation"},
        "conditionExpressionType": {"naturalLanguage", "feel", "javascript", "sql"},
        "testScenarioType": {"normal", "exception", "boundary", "compensation"},
    },
    "messageFlow": {
        "deliveryMode": {"sync", "async"},
        "testScenarioType": {"normal", "exception", "boundary", "compensation"},
    },
    "association": {
        "direction": {"none", "oneWay", "twoWay"},
        "dataRole": {"input", "output", "reference"},
        "testScenarioType": {"normal", "exception", "boundary", "compensation"},
    },
}

FIELD_LABELS = {
    "triggerType": "触发方式", "startCondition": "启动条件",
    "resultType": "结果类型", "finalBusinessState": "最终业务状态",
    "transactionType": "事务类型", "decisionVariable": "判断变量",
    "sourceType": "来源类型", "targetType": "目标类型",
    "storeType": "存储类型", "flowKind": "路径类型",
}


def _text(value: Any) -> str:
    return str(value).strip() if value is not None else ""


def _items(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []


def _text_list(value: Any) -> list[str]:
    if isinstance(value, str):
        return [item.strip() for item in value.splitlines() if item.strip()]
    return [_text(item) for item in _items(value) if _text(item)]


def _objects(value: Any) -> list[dict[str, Any]]:
    return [dict(item) for item in _items(value) if isinstance(item, Mapping)]


def _profile_value(source: Mapping[str, Any], snake: str, camel: str) -> str | None:
    value = _text(source.get(snake) or source.get(camel))
    return value or None


def node_semantic_type(node: Mapping[str, Any]) -> str | None:
    if _profile_value(node, "bpmn_element_type", "bpmnElementType") == "TASK":
        return None
    key = (
        _profile_value(node, "bpmn_element_type", "bpmnElementType"),
        _profile_value(node, "bpmn_event_kind", "bpmnEventKind"),
        _profile_value(node, "bpmn_gateway_type", "bpmnGatewayType"),
        _profile_value(node, "bpmn_subprocess_kind", "bpmnSubProcessKind"),
    )
    return NODE_SEMANTIC_TYPES.get(key)


def edge_semantic_type(edge: Mapping[str, Any]) -> str | None:
    profile = (
        _profile_value(edge, "bpmn_flow_type", "bpmnFlowType")
        or _profile_value(edge, "edge_type", "edgeType")
        or ""
    )
    return EDGE_SEMANTIC_TYPES.get(profile)


def bpmn_semantic_payload(
    value: Any,
    semantic_type: str | None,
    fallback_name: str | None = None,
) -> dict[str, Any]:
    if not semantic_type or semantic_type not in FIELD_KINDS:
        return {}
    source = value if isinstance(value, Mapping) else {}
    result: dict[str, Any] = {
        "schemaVersion": SCHEMA_VERSION,
        "semanticType": semantic_type,
    }
    defaults = DEFAULT_VALUES.get(semantic_type, {})
    for key, kind in FIELD_KINDS[semantic_type].items():
        raw = source.get(key)
        if kind == "text":
            text = _text(raw) or _text(defaults.get(key))
            allowed = ENUM_VALUES.get(semantic_type, {}).get(key)
            result[key] = text if not allowed or text in allowed else _text(defaults.get(key))
        elif kind == "list":
            result[key] = _text_list(raw)
        elif kind == "objects":
            result[key] = [
                {
                    "flowId": _text(item.get("flowId")),
                    "label": _text(item.get("label")),
                    "condition": _text(item.get("condition")),
                }
                for item in _objects(raw)
            ]
        elif kind == "bool":
            result[key] = raw if isinstance(raw, bool) else bool(defaults.get(key, False))
        elif kind == "int":
            try:
                result[key] = int(raw) if raw not in (None, "") else defaults.get(key)
            except (TypeError, ValueError):
                result[key] = defaults.get(key)
    name_key = NAME_KEYS[semantic_type]
    if not _text(result.get(name_key)):
        result[name_key] = _text(fallback_name)
    return result


def semantic_display_name(payload: Mapping[str, Any], fallback: str | None = None) -> str:
    name_key = NAME_KEYS.get(_text(payload.get("semanticType")))
    value = _text(payload.get(name_key)) if name_key else ""
    return value or _text(fallback)


def bpmn_semantic_quality_issues(
    element: Mapping[str, Any],
    payload: Mapping[str, Any],
    target: str,
) -> list[dict[str, Any]]:
    semantic_type = node_semantic_type(element) if target == "NODE" else edge_semantic_type(element)
    if not semantic_type:
        return []
    key = _text(element.get("node_key") or element.get("edge_key"))
    normalized = bpmn_semantic_payload(
        payload,
        semantic_type,
        _text(element.get("title") or element.get("label")),
    )
    issues: list[dict[str, Any]] = []

    def add(code: str, message: str) -> None:
        issues.append({
            "targetType": target,
            "targetKey": key,
            "severity": "warning",
            "code": code,
            "message": message,
        })

    if not semantic_display_name(normalized):
        add("BPMN_SEMANTIC_MISSING_NAME", "元素缺少业务名称。")
    required = {
        "startEvent": ("triggerType", "startCondition"),
        "endEvent": ("resultType", "finalBusinessState"),
        "transaction": ("transactionType",),
        "exclusiveGateway": ("decisionVariable",),
        "dataInput": ("sourceType",),
        "dataOutput": ("targetType",),
        "dataStore": ("storeType",),
        "sequenceFlow": ("flowKind",),
    }.get(semantic_type, ())
    for field in required:
        if not normalized.get(field):
            add("BPMN_SEMANTIC_MISSING_FIELD", f"元素缺少必填业务字段：{FIELD_LABELS.get(field, field)}。")
    if (
        semantic_type == "transaction"
        and normalized.get("transactionType") == "saga"
        and (
            not normalized.get("cancelTriggers")
            or not normalized.get("compensationPolicy")
            or not normalized.get("compensationTasks")
        )
    ):
        add("BPMN_SEMANTIC_SAGA_MISSING_COMPENSATION", "Saga 事务缺少取消条件或补偿定义。")
    if semantic_type == "complexGateway":
        required_count = normalized.get("requiredCount")
        total_count = normalized.get("totalCount")
        if required_count and total_count and required_count > total_count:
            add("BPMN_SEMANTIC_INVALID_COUNT", "复杂网关要求完成数量不能大于总分支数。")
    if semantic_type in {"exclusiveGateway", "inclusiveGateway"}:
        branch_key = "branches" if semantic_type == "exclusiveGateway" else "branchConditions"
        flow_ids = [
            _text(branch.get("flowId"))
            for branch in normalized.get(branch_key, [])
            if _text(branch.get("flowId"))
        ]
        if len(flow_ids) != len(set(flow_ids)):
            add("BPMN_SEMANTIC_DUPLICATE_BRANCH", "网关存在重复的分支连线。")
        if (
            semantic_type == "exclusiveGateway"
            and normalized.get("defaultFlowId")
            and normalized["defaultFlowId"] not in flow_ids
        ):
            add("BPMN_SEMANTIC_INVALID_DEFAULT_FLOW", "默认路径不在已配置分支中。")
    return issues


def is_data_element(element: Mapping[str, Any]) -> bool:
    return _profile_value(element, "bpmn_element_type", "bpmnElementType") in DATA_ELEMENT_TYPES

File: front/src/entities/business-flow/model/bpmnSemantic.ts
import type {
  BpmnEdgeProfile,
  BpmnNodeProfile,
  BpmnSemanticJson,
  BpmnSemanticType,
} from './types'

const DATA_TYPES = new Set(['DATA_OBJECT', 'DATA_INPUT', 'DATA_OUTPUT', 'DATA_STORE'])

export function isDataBpmnProfile(profile: BpmnNodeProfile) {
  return DATA_TYPES.has(profile.bpmnElementType)
}

export function nodeSemanticType(profile: BpmnNodeProfile): BpmnSemanticType | null {
  if (profile.bpmnElementType === 'TASK') return null
  if (profile.bpmnElementType === 'EVENT') {
    if (profile.bpmnEventKind === 'START') return 'startEvent'
    if (profile.bpmnEventKind === 'END') return 'endEvent'
    return 'intermediateEvent'
  }
  if (profile.bpmnElementType === 'SUB_PROCESS' && profile.bpmnSubProcessKind === 'TRANSACTION') {
    return 'transaction'
  }
  if (profile.bpmnElementType === 'GATEWAY') {
    if (profile.bpmnGatewayType === 'INCLUSIVE') return 'inclusiveGateway'
    if (profile.bpmnGatewayType === 'PARALLEL') return 'parallelGateway'
    if (profile.bpmnGatewayType === 'COMPLEX') return 'complexGateway'
    return 'exclusiveGateway'
  }
  if (profile.bpmnElementType === 'DATA_OBJECT') return 'dataObject'
  if (profile.bpmnElementType === 'DATA_INPUT') return 'dataInput'
  if (profile.bpmnElementType === 'DATA_OUTPUT') return 'dataOutput'
  if (profile.bpmnElementType === 'DATA_STORE') return 'dataStore'
  return null
}

export function edgeSemanticType(profile: BpmnEdgeProfile): BpmnSemanticType {
  if (profile.bpmnFlowType === 'MESSAGE') return 'messageFlow'
  if (profile.bpmnFlowType === 'ASSOCIATION') return 'association'
  return 'sequenceFlow'
}

const NAME_KEYS: Record<BpmnSemanticType, string> = {
  startEvent: 'eventName',
  intermediateEvent: 'eventName',
  endEvent: 'eventName',
  transaction: 'transactionName',
  exclusiveGateway: 'decisionName',
  inclusiveGateway: 'decisionName',
  parallelGateway: 'gatewayName',
  complexGateway: 'gatewayName',
  dataObject: 'dataName',
  dataInput: 'dataName',
  dataOutput: 'dataName',
  dataStore: 'dataName',
  sequenceFlow: 'flowName',
  messageFlow: 'messageName',
  association: 'associationName',
}

export const BPMN_SEMANTIC_TYPES = Object.keys(NAME_KEYS) as BpmnSemanticType[]

const ENUM_VALUES: Partial<Record<BpmnSemanticType, Record<string, string[]>>> = {
  startEvent: { triggerType: ['manual', 'message', 'timer', 'signal', 'condition'] },
  intermediateEvent: {
    catchOrThrow: ['catch', 'throw'],
    eventDefinition: ['message', 'timer', 'error', 'signal', 'compensation', 'escalation'],
  },
  endEvent: { resultType: ['success', 'failure', 'cancel', 'terminate', 'error', 'compensation'] },
  transaction: {
    transactionType: ['technical', 'business', 'saga'],
    compensationOrder: ['reverseOrder', 'forwardOrder', 'custom'],
    consistencyLevel: ['strong', 'eventual'],
  },
  exclusiveGateway: { conditionExpressionType: ['naturalLanguage', 'feel', 'javascript', 'sql'] },
  parallelGateway: { parallelMode: ['fork', 'join', 'forkJoin'] },
  dataInput: {
    sourceType: ['user', 'system', 'api', 'message', 'file'],
    sensitiveLevel: ['normal', 'internal', 'confidential'],
  },
  dataOutput: { targetType: ['api', 'database', 'message', 'file', 'ui'] },
  dataStore: {
    storeType: ['database', 'cache', 'file', 'mq', 'externalSystem'],
    accessMode: ['read', 'write', 'readWrite'],
    consistencyLevel: ['strong', 'eventual'],
    privacyLevel: ['normal', 'internal', 'confidential'],
  },
  sequenceFlow: {
    flowKind: ['normal', 'conditional', 'default', 'exception', 'compensation'],
    conditionExpressionType: ['naturalLanguage', 'feel', 'javascript', 'sql'],
    testScenarioType: ['normal', 'exception', 'boundary', 'compensation'],
  },
  messageFlow: {
    deliveryMode: ['sync', 'async'],
    testScenarioType: ['normal', 'exception', 'boundary', 'compensation'],
  },
  association: {
    direction: ['none', 'oneWay', 'twoWay'],
    dataRole: ['input', 'output', 'reference'],
    testScenarioType: ['normal', 'exception', 'boundary', 'compensation'],
  },
}

const EMPTY_BY_TYPE: Record<BpmnSemanticType, Record<string, unknown>> = {
  startEvent: { eventName: '', triggerType: 'manual', triggerSource: '', startCondition: '', inputDataRefs: [], initiator: '', frequency: '', preCheckRules: [] },
  intermediateEvent: { eventName: '', catchOrThrow: 'catch', eventDefinition: 'message', interrupting: false, timeout: '', messageName: '', errorCode: '', escalationCode: '', businessMeaning: '' },
  endEvent: { eventName: '', resultType: 'success', finalBusinessState: '', outputDataRefs: [], notifyTargets: [], auditRequired: false, rollbackRequired: false },
  transaction: { transactionName: '', transactionType: 'business', successCriteria: [], cancelTriggers: [], compensationPolicy: '', compensationOrder: 'reverseOrder', consistencyLevel: 'eventual', timeout: '', isolationNote: '', compensationTasks: [], partialSuccessPolicy: '', auditRequired: false },
  exclusiveGateway: { decisionName: '', decisionVariable: '', branches: [], defaultFlowId: '', conditionExpressionType: 'naturalLanguage', mutuallyExclusive: true, coverageRequired: true },
  inclusiveGateway: { decisionName: '', branchConditions: [], allowMultipleBranches: true, mergePolicy: '', minSelectedBranches: 1, coverageRequired: true },
  parallelGateway: { gatewayName: '', parallelMode: 'forkJoin', waitForAll: true, expectedBranches: [], partialFailurePolicy: '', timeout: '', concurrencyLimit: null },
  complexGateway: { gatewayName: '', activationCondition: '', completionCondition: '', requiredCount: null, totalCount: null, customRule: '', explanation: '' },
  dataObject: { dataName: '', entityName: '', schemaRef: '', lifecycleState: '', ownerActivityRef: '', readByRefs: [], writeByRefs: [] },
  dataInput: { dataName: '', sourceType: 'user', sourceRef: '', required: false, validationRules: [], exampleValue: '', sensitiveLevel: 'normal', defaultValue: '' },
  dataOutput: { dataName: '', targetType: 'ui', targetRef: '', outputContract: '', transformRule: '', successOutput: '', failureOutput: '' },
  dataStore: { dataName: '', storeType: 'database', systemRef: '', accessMode: 'readWrite', consistencyLevel: 'eventual', retentionPolicy: '', privacyLevel: 'internal' },
  sequenceFlow: { flowName: '', flowKind: 'normal', conditionText: '', conditionExpression: '', conditionExpressionType: 'naturalLanguage', priority: null, isDefault: false, businessRuleRefs: [], testScenarioType: 'normal', expectedResult: '' },
  messageFlow: { messageName: '', businessMeaning: '', senderRef: '', receiverRef: '', payloadDataRefs: [], deliveryMode: 'async', timeout: '', testScenarioType: 'normal', expectedResult: '' },
  association: { associationName: '', businessMeaning: '', direction: 'none', dataRole: 'reference', testScenarioType: 'normal', expectedResult: '' },
}

export function emptyBpmnSemantic(type: BpmnSemanticType, name = ''): BpmnSemanticJson {
  return {
    schemaVersion: 1,
    semanticType: type,
    ...EMPTY_BY_TYPE[type],
    [NAME_KEYS[type]]: name,
  } as BpmnSemanticJson
}

export function normalizeBpmnSemantic(
  value: unknown,
  type: BpmnSemanticType,
  fallbackName = '',
): BpmnSemanticJson {
  const source = value && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {}
  const allowed = EMPTY_BY_TYPE[type]
  const normalized: Record<string, unknown> = { schemaVersion: 1, semanticType: type }
  Object.entries(allowed).forEach(([key, fallback]) => {
    const raw = source[key]
    if (key === 'branches' || key === 'branchConditions') {
      normalized[key] = Array.isArray(raw)
        ? raw.filter((item) => item && typeof item === 'object' && !Array.isArray(item)).map((item) => {
            const branch = item as Record<string, unknown>
            return {
              flowId: typeof branch.flowId === 'string' ? branch.flowId.trim() : '',
              label: typeof branch.label === 'string' ? branch.label.trim() : '',
              condition: typeof branch.condition === 'string' ? branch.condition.trim() : '',
            }
          })
        : []
    } else if (Array.isArray(fallback)) {
      normalized[key] = Array.isArray(raw)
        ? raw.filter((item): item is string => typeof item === 'string').map((item) => item.trim())
        : []
    }
    else if (typeof fallback === 'boolean') normalized[key] = typeof raw === 'boolean' ? raw : fallback
    else if (typeof fallback === 'number' || fallback === null) {
      const number = Number(raw)
      normalized[key] = raw === '' || raw === undefined || raw === null || !Number.isFinite(number)
        ? fallback
        : number
    } else {
      const fallbackText = typeof fallback === 'string' ? fallback : ''
      const text = typeof raw === 'string' ? raw.trim() : fallbackText
      const allowedValues = ENUM_VALUES[type]?.[key]
      normalized[key] = allowedValues && !allowedValues.includes(text) ? fallbackText : text
    }
  })
  const nameKey = NAME_KEYS[type]
  normalized[nameKey] = String(normalized[nameKey] || fallbackName).trim()
  return normalized as BpmnSemanticJson
}

export function bpmnSemanticDisplayName(value: BpmnSemanticJson | null | undefined, fallback = '') {
  if (!value) return fallback
  return String((value as unknown as Record<string, unknown>)[NAME_KEYS[value.semanticType]] || fallback).trim()
}

export function bpmnSemanticQualityIssues(value: BpmnSemanticJson): string[] {
  const issues: string[] = []
  if (!bpmnSemanticDisplayName(value)) issues.push('缺少业务名称')
  const data = value as unknown as Record<string, unknown>
  if (value.semanticType === 'startEvent' && (!data.triggerType || !data.startCondition)) issues.push('开始事件缺少触发方式或启动条件')
  if (value.semanticType === 'endEvent' && (!data.resultType || !data.finalBusinessState)) issues.push('结束事件缺少结果类型或最终业务状态')
  if (value.semanticType === 'transaction' && data.transactionType === 'saga') {
    if (!(data.cancelTriggers as unknown[])?.length || !data.compensationPolicy || !(data.compensationTasks as unknown[])?.length) issues.push('Saga 事务缺少取消条件或补偿定义')
  }
  if (value.semanticType === 'exclusiveGateway' && !data.decisionVariable) issues.push('排他网关缺少判断变量')
  if (value.semanticType === 'exclusiveGateway' || value.semanticType === 'inclusiveGateway') {
    const branches = (data.branches ?? data.branchConditions) as { flowId?: string }[]
    const ids = branches.map((branch) => branch.flowId).filter(Boolean) as string[]
    if (new Set(ids).size !== ids.length) issues.push('网关存在重复的分支连线')
    if (value.semanticType === 'exclusiveGateway' && data.defaultFlowId && !ids.includes(String(data.defaultFlowId))) {
      issues.push('默认路径不在已配置分支中')
    }
  }
  if (value.semanticType === 'complexGateway') {
    const required = data.requiredCount as number | null
    const total = data.totalCount as number | null
    if (required !== null && total !== null && required > total) issues.push('要求完成数量不能大于总分支数量')
  }
  return issues
}

export const BPMN_SEMANTIC_ENUM_LABELS: Record<string, string> = {
  manual: '人工触发', message: '消息触发', timer: '定时触发', signal: '信号触发',
  condition: '条件触发', catch: '捕获', throw: '抛出', error: '错误',
  compensation: '补偿', escalation: '升级', success: '成功', failure: '失败',
  cancel: '取消', terminate: '终止', technical: '技术事务', business: '业务事务',
  saga: '长事务补偿', reverseOrder: '逆序补偿', forwardOrder: '顺序补偿',
  custom: '自定义', strong: '强一致', eventual: '最终一致',
  naturalLanguage: '自然语言', feel: '规则表达式',
  javascript: '脚本表达式', sql: '查询表达式', fork: '并行拆分', join: '并行汇聚',
  forkJoin: '拆分并汇聚', user: '用户', system: '系统', api: '接口',
  file: '文件', database: '数据库', cache: '缓存', mq: '消息队列',
  externalSystem: '外部系统', ui: '页面', read: '只读', write: '只写',
  readWrite: '读写', normal: '普通', conditional: '条件', default: '默认',
  exception: '异常', boundary: '边界', async: '异步', sync: '同步', none: '无方向',
  oneWay: '单向', twoWay: '双向', input: '输入', output: '输出',
  reference: '引用', internal: '内部', confidential: '机密',
}
